### PR TITLE
ledger-configuration: Extract the configuration from participant-state. [KVL-1002]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
The configuration is often used without the state, and it probably doesn't need to be versioned in the same way.

This is a big PR. Apologies for that. It's pretty much entirely renames and repackaging; no behavior has changed except that we now publish an extra Maven package, _ledger-configuration_.

### Changelog

- **[Integration Kit]** The ledger configuration classes, `Configuration`, `LedgerInitialConditions`, and `TimeModel`, have been moved from _participant-state_ to a separate package named _ledger-configuration_, in the Java package `com.daml.ledger.configuration`. You will need to update your dependencies and imports.
- **[Integration Kit]** `TimeModel` has been renamed to `LedgerTimeModel`. If you are using the ledger configuration classes directly, you may need to update your code.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by SamirTalwar-DA** [comment]
@gerolf-da, would you mind chiming in on whether this is a good idea or not?

**Comment by SamirTalwar-DA** [comment]
To add a bit more context: the purpose of this is to decouple the _participant-state-index_ package from _participant-state_, so we can use v1 and v2 of the participant state interchangeably.

If this is acceptable, I will follow up with PRs to move `Offset` into a new package, and use the Daml-LF `Ref` object directly instead of the type aliases in _v1/package.scala_.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/AdaptedV1ReadService.scala
Comment (hubert-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/AdaptedV1ReadService.scala:36]: 👍🏻

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/AdaptedV1WriteService.scala
Comment (hubert-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/AdaptedV1WriteService.scala:167]: 👍🏻

**Review by hubert-da — APPROVED** [review]
LGTM at a low level. You may want to wait for a higher level approval ;)

**Review by gerolf-da — CHANGES_REQUESTED** [review]
@SamirTalwar-DA: I'd like to understand the motivation for this change a bit better, please.

**Review by miklos-da — APPROVED** [review]

**Comment by SamirTalwar-DA** [comment]
@gerolf-da: Happy to discuss any concerns here or over a call. :smiley:

File: ledger/ledger-configuration/protobuf/ledger_configuration.rst
Comment (fabiotudone-da) [reply ledger/ledger-configuration/protobuf/ledger_configuration.rst:23]: I'd check usages and, if we can safely do it, I think it's better to do it (not necessarily as part of this PR).

**Review by fabiotudone-da — COMMENTED** [review]
I am not so confident about the ledger time model not having the same versioning model of the API, as it is (also) part of it; anyway, supposing we merge this change and then we need to start versioning it, how would you do it?

**Comment by SamirTalwar-DA** [comment]
@fabiotudone-da: The protobuf, _ledger_configuration.proto_, is versioned, which I think takes care of that problem. The classes can (I believe) be evolved safely, as long as they're always able to decode later versions of the protobuf.

Take a look at _Configuration.scala_. It has two `decode` methods, where v1 provides a sensible default for `maxDeduplicationTime`.

Do you think this satisfies your concern?

**Review by rautenrieth-da — COMMENTED** [review]
I had a quick high-level look at these changes.

My only concern was that now two different versions of the API (v1 and v2) depend on the same types, which means we'll have to be careful when changing those types (e.g., if `TimeModel` should change in v2 but not in v1). After discussion with @SamirTalwar-DA, I do not think this is an issue because:

1. We do not expect changes to the time model and the configuration
2. The v1 and v2 versions behave more like the stable and preview version of "the" participant state API, the main versioning of which is the SDK version.

**Comment by fabiotudone-da** [comment]
> I had a quick high-level look at these changes.
> 
> My only concern was that now two different versions of the API (v1 and v2) depend on the same types, which means we'll have to be careful when changing those types (e.g., if `TimeModel` should change in v2 but not in v1). After discussion with @SamirTalwar-DA, I do not think this is an issue because:
> 
> 1. We do not expect changes to the time model and the configuration
> 2. The v1 and v2 versions behave more like the stable and preview version of "the" participant state API, the main versioning of which is the SDK version.

@SamirTalwar-DA @rautenrieth-da I think this satisfies my concerns. Thanks!

**Review by fabiotudone-da — APPROVED** [review]

**Comment by SamirTalwar-DA** [comment]
OK, I'm going to merge this, with a bit of trepidation.

Worst-case scenario, we revert.

File: ledger/ledger-configuration/protobuf/ledger_configuration.rst
Comment (SamirTalwar-DA) [reply ledger/ledger-configuration/protobuf/ledger_configuration.rst:23]: #10302 covers this.

**Pull request merged**
